src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -3082,8 +3082,20 @@ private IEnumerable<PSModuleInfo> CreateFakeModuleObject(IEnumerable<ModuleSpeci
                 // In that case, the nested module will first be loaded with a different session state, and then when trying to load the RootModule via 'LoadModuleNamedInManifest',
                 // the same loaded nested module will be reused for the RootModule by 'LoadModuleNamedInManifest'.
 
-                // Change the module name to match the manifest name, not the original name
+                // Initially the original module name comes from RootModule manifest property but real name comes from the module (manifest) path.
+                // Change the module name to match the manifest name, not the original name.
+                // Then do the same for all providers loaded in the module since a provider caches full name based on the module name.
                 newManifestInfo.SetName(manifestInfo.Name);
+                if (ss != null)
+                {
+                    foreach (ProviderInfo pi in ss.Provider.GetAll())
+                    {
+                        if (object.ReferenceEquals(pi.Module, newManifestInfo))
+                        {
+                            pi.SetModule(manifestInfo);
```

Comment:
The `pi.Module` is exactly the same object as `newManifestInfo`, but you are calling `SetModule` again. So it looks to me you are trying to use the side effect of the `SetModule` call, which set `_fullName = null`, right?
This is quite obscure, and it's very easy to be removed by a code cleanup work because it looks like unnecessary code. If the right fix is to set `_fullName = null` here, then maybe it's better to just add a setter to the `FullName` property of `ProviderInfo`.

I think this PR is not ready to be merged, we can include this PR in the next servicing release of PS 7.

Reply:
> This is quite obscure, and it's very easy to be removed by a code cleanup work because it looks like unnecessary code.

Why does it look like unnecessary code? Also if anybody remove the code our test will fail. And comment above explicitly says that we do here.

Reply:
> Why does it look like unnecessary code?

`pi.Module` and `newManifestInfo` are exactly the same object already when getting in the if block, and thus calling `SetModule(manifestInfo)` to set the exact same `ModuleInfo` object seems unneeded.
Like I said, if the right fix is to set `_fullName = null` (_which I'm not sure as I don't fully understand the bug yet_), then it's much more readable to just add a setter to `FullName` property.

Reply:
:-) It seems we look on the same thing with different sides.

`FullName` is a _computed_ property. It is calculated from the module name and the provider short name. So directly set `FullName` is not right way I believe.

Reply:
@daxian-dbw - do you agree?